File: infer/tests/codetoanalyze/objc/pulse/use_after_free.m
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#import <Foundation/NSObject.h>

@interface PulseTest : NSObject

- (int)use_after_free_simple_in_objc_method_bad:(int*)x;

@end

@implementation PulseTest

- (int)use_after_free_simple_in_objc_method_bad:(int*)x {
  free(x);
  return *x;
}

@end

int use_after_free_simple_bad(int* x) {
  free(x);
  return *x;
}

File: infer/tests/build_systems/pulse_messages_c/use_after_free.c
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include <stdlib.h>

void uaf_intraprocedural_bad(int* x) {
  free(x);
  int _ = *x;
}

void free_wrapper(int* p) { free(p); }

void free_wrapper_uaf_bad(int* x) {
  free_wrapper(x);
  int _ = *x;
}

void deref_int(int* x) { int _ = *x; }

void uaf_via_deref_bad(int* x) {
  free(x);
  deref_int(x);
}

void uaf_via_deref_and_free_wrapper_bad(int* x) {
  free_wrapper(x);
  deref_int(x);
}
